framework/src/base/NonlinearSystem.C

```diff
@@ -213,6 +214,45 @@ NonlinearSystem::stopSolve()
 void
 NonlinearSystem::setupFiniteDifferencedPreconditioner()
 {
+  std::shared_ptr<FiniteDifferencePreconditioner> fdp =
+      std::dynamic_pointer_cast<FiniteDifferencePreconditioner>(_preconditioner);
+  if (!fdp)
+    mooseError("Did not setup finite difference preconditioner, and please add a preconditioning "
+               "block with type = fdp");
+
+  if (fdp->finiteDifferenceType() == "coloring")
+    setupColoringFiniteDifferencedPreconditioner();
+  else if (fdp->finiteDifferenceType() == "standard")
+    setupStandardFiniteDifferencedPreconditioner();
+  else
+    mooseError("Unknown finite difference type");
+}
+
+void
+NonlinearSystem::setupStandardFiniteDifferencedPreconditioner()
+{
+#if LIBMESH_HAVE_PETSC
+  PetscNonlinearSolver<Number> * petsc_nonlinear_solver =
+      dynamic_cast<PetscNonlinearSolver<Number> *>(_transient_sys.nonlinear_solver.get());
+
+  if (!petsc_nonlinear_solver)
+    mooseError("Did not have a nonlinear solver setup in libMesh yet");
+
+  PetscMatrix<Number> * petsc_mat = dynamic_cast<PetscMatrix<Number> *>(_transient_sys.matrix);
+  if (!petsc_mat)
+    mooseError("Transient system does not have a PETSc mat");
```

Comment:
If these new functions are made private, you might be able to just go to a static cast here because you'll know when they are called. I'm not sure I see the utility in making them public. Will they ever be used outside of the class?